kernel/fatfs.c

```diff
@@ -133,10 +134,37 @@ STATIC void init_direntry(struct dirent *dentry, unsigned attrib,
 int dos_open(char *path, unsigned flags, unsigned attrib, int fd)
 {
   REG f_node_ptr fnp = sft_to_fnode(fd);
-  int status = find_fname(path, D_ALL | attrib, fnp);
+  int status;
+
+  /* Special handling for volume name (they can coexist with files/directories)
+   */
+  if (((attrib & (D_DIR | D_VOLID | D_SYSTEM | D_HIDDEN | D_RDONLY)) == D_VOLID) && (flags & O_TRUNC)) {
+    int ret;
+    char rpath[] = "?:\\";
+
+    /* Count the slashes to see if we are in the root directory */
+    if (strchr(path + 3, '\\'))
+      return DE_ACCESS;
+
+    /* Check to see if there are any other files with VOLID set */
+    rpath[0] = path[0];
```

Comment:
What do we do here if the
initial path is not in a root dir?
I think you need to look into
the same dir, not in root dir,
as you did before. But I don't
know what VOL files do in a
non-root dir.

Reply:
> What do we do here if the
initial path is not in a root dir?

Nothing because the code I added above it returns DE_ACCESS if we are not in the root directory.

> But I don't
know what VOL files do in a
non-root dir.

See what you make of this https://fd.lod.bz/rbil/interrup/dos_kernel/2111.html? I'm not sure quite what to think, does it mean that if asking for file with vol attr set in a non root directory, it returns any that it finds even though they don't really mean anything when not in the root, or does it mean it returns the volume file from the root?

Reply:
Not sure what FindFirst returns,
and not sure its important here.
Here we _create_ the file with
vol attr. So some other rules
should apply. If you have tested
that any such access returns
DE_ACCESS on any DOS then
fine.

Reply:
> Here we create the file with
vol attr. So some other rules
should apply. If you have tested
that any such access returns
DE_ACCESS on any DOS then
fine.

Yes I should do that.

Reply:
Well, not what I expected. It seems that MS-DOS 6.22 and DR-DOS 7.01 see the attempt to create a file with volume ID bit set in a non root directory and actually write it in the root, so I have to change the patch.